### PR TITLE
Support Indexing of the SAM template.

### PR DESCRIPTION
**Pull request by zhangzhx**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

Add indexing for SAM template file to index resources. The key is the resource type, and the value is a collection of the resources of the type. Only a subset of the resource properties are persisted for the indexing such as Function runtime and handler.

**TODO:**
1. Add a setting for enabling showing all the valid handlers
1. Add more tests for all the edge cases of showing the line marker

## Related Issue(s)

#309

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            }
+        }
+
+        private fun listAllResources(project: Project): Collection<IndexedResource> {
+            val index = FileBasedIndex.getInstance()
+            return index.getAllKeys(NAME, project)
+                    .asSequence()
+                    .mapNotNull { index.getValues(NAME, it, GlobalSearchScope.projectScope(project)) }
+                    .filter { it.isNotEmpty() }
+                    .flatten()
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt]: The `index.getValues` returns a list of `Value`s. In this case, the `Value` type is already a list of resources. That's why I flatten twice since it is kinda 2d array.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt [hunk 2]
```diff
@@ -64,6 +72,11 @@ class LambdaLineMarker : LineMarkerProviderDescriptor() {
     ) {
     }
 
+    private fun isHandlerValid(element: PsiElement, handler: String, runtime: Runtime): Boolean =
+        !handler.contains("::") || // Class level handler is valid
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt]: Considering handler **NOT** containing `::` valid. **NOT** containing `::` is regarded as a class handler which means it implements the Lambda interface. It is valid no matter it is in template or remote.

[the remark this answers is not in the document]
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt]: Python/NodeJS/PowerShell/Go never contain `::` though

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt [hunk 3]
```diff
@@ -0,0 +1,54 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.cloudformation
+
+/**
+ * Immutable data class for indexing [Resource]. Use [from] or [fromResource] to create an instance so that it always
+ * returns a concrete [IndexedResource] such as [IndexedFunction] if applicable.
+ */
+open class IndexedResource protected constructor(val logicalName: String, val type: String, val indexedProperties: Map<String, String>) {
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: Now we only index Lambda `runtime` and `handler`. We are not indexing properties of other resources in the way that if the resource is not modeled, ie not in the `INDEXED_RESOURCE_MAP`, we just keep the logical names. I don't think that costs much.

Thoughts?

[the remark this answers is not in the document]
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: To what gain? If we don't use it, we shouldn't have the code. It's just more code to maintain.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt]: Ah

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                dataOutput.writeInt(resource.indexedProperties.size)
+                resource.indexedProperties.forEach { key, value ->
+                    dataOutput.writeUTF(key)
+                    dataOutput.writeUTF(value)
+                }
+            }
+        }
+
+        override fun read(dataInput: DataInput): MutableList<IndexedResource> {
+            val size = dataInput.readInt()
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt]: can we call this `resourceCount` - I know it is literally `Collection.size` but it represents the number of resources being read

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+    override fun getIndexer(): DataIndexer<String, MutableList<IndexedResource>, FileContent> = DataIndexer { fileContent ->
+        val indexedResources = mutableMapOf<String, MutableList<IndexedResource>>()
+
+        fileContent.psiFile.acceptNode(object : PsiElementVisitor() {
+            override fun visitElement(element: PsiElement?) {
+                super.visitElement(element)
+                element?.run {
+                    val parent = element.parent as? YAMLKeyValue ?: return
+                    if (parent.value != this) return
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt]: what is this guarding against?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                    .flatten()
+                    .flatten()
+                    .toList()
+        }
+
+        fun listResourcesByType(project: Project, type: String): Collection<IndexedResource> = listAllResources(project) { it == type }
+
+        fun listFunctions(project: Project): Collection<IndexedFunction> =
+                listAllResources(project) { it == SERVERLESS_FUNCTION_TYPE || it == LAMBDA_FUNCTION_TYPE }
+                        .asSequence()
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt]: Why do `asSequence` if we're just going to convert it immediately to a list?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt [hunk 7]
```diff
@@ -37,25 +42,28 @@ class LambdaLineMarker : LineMarkerProviderDescriptor() {
...
-        val actionGroup = DefaultActionGroup()
+        return if (isHandlerValid(element, handler, runtime)) {
+            val actionGroup = DefaultActionGroup()
 
-        if (element.language in LambdaPackager.supportedLanguages) {
-            val executorActions = ExecutorAction.getActions(1)
-            executorActions.forEach {
-                actionGroup.add(LineMarkerActionWrapper(element, it))
-            }
+            if (element.language in LambdaPackager.supportedLanguages) {
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt]: Something weird going on with your spacing here - can you check that you're using the same style conventions?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt [hunk 8]
```diff
@@ -64,6 +72,11 @@ class LambdaLineMarker : LineMarkerProviderDescriptor() {
     ) {
     }
 
+    private fun isHandlerValid(element: PsiElement, handler: String, runtime: Runtime): Boolean =
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt]: It's not really `isValid` is more like `shouldShowLineMarker`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt [hunk 9]
```diff
@@ -64,6 +72,11 @@ class LambdaLineMarker : LineMarkerProviderDescriptor() {
...
+        (runtime.runtimeGroup == RuntimeGroup.JAVA && !handler.contains("::")) || // Java class level handler is valid
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt]: We should push this language specific piece down to the Java extension point; which would give other runtimes a hook too which will enable them to override when the line marker should be shown.

Go for example would probably always want to show the line markers because the requirements for them are extremely specific to Lambda - and so it's obviously a Lambda.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    protected constructor(resource: Resource, indexProperties: List<String>)
+            : this(resource.logicalName, resource.type()!!, indexProperties.map { it to resource.getScalarProperty(it) }.toMap())
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: Fixed

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+                super.visitElement(element)
+                element?.run {
+                    val parent = element.parent as? YAMLKeyValue ?: return
+                    if (parent.value != this) return
+
+                    val indexedResource = YamlCloudFormationTemplate.convertPsiToResource(parent)?.let {
+                        IndexedResource.fromResource(it)
+                    } ?: return
+
+                    (indexedResources.computeIfAbsent(indexedResource.type) { mutableListOf() } as MutableList)
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt]: Fixed

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt]: Moved the language specific condition to `LambdaHandlerResolver`

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: Removed `logicalName` and `type` of indexing

File: jetbrains-core/resources/META-INF/plugin.xml [hunk 12]
```diff
@@ -83,6 +83,8 @@
         <programRunner implementation="software.aws.toolkits.jetbrains.services.lambda.execution.sam.SamInvokeRunner"/>
         <programRunner implementation="software.aws.toolkits.jetbrains.services.lambda.execution.remote.RemoteLambdaRunner"/>
         <fileBasedIndex implementation="software.aws.toolkits.jetbrains.services.lambda.LambdaHandlerIndex"/>
+        <fileBasedIndex implementation="software.aws.toolkits.jetbrains.services.lambda.SamTemplateIndex"/>
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/resources/META-INF/plugin.xml]: Make sense. We'll index other resources in CFN as well in the future.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        return true
+    }
+
+    override fun hashCode(): Int = indexedProperties.hashCode()
+
+    companion object {
+        fun from(type: String, indexedProperties: Map<String, String>) =
+                INDEXED_RESOURCE_MAPPINGS[type]?.first?.invoke(indexedProperties) ?: IndexedResource(indexedProperties)
+
+        fun fromResource(resource: Resource): IndexedResource? =
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: done.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+ * returns a concrete [IndexedResource] such as [IndexedFunction] if applicable.
+ */
+open class IndexedResource internal constructor(val indexedProperties: Map<String, String>) {
+
+    protected constructor(resource: Resource, indexProperties: List<String>)
+            : this(indexProperties.map { it to resource.getScalarProperty(it) }.toMap())
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is IndexedResource) return false
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: This was auto generated. Made the change.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt (hunk 4 above)
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt]: done.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt (hunk 5 above)
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt]: This makes sure we only stop at the `YAMLKeyValue.value` element for this visit. The parent could have other children which we don't care.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt (hunk 6 above)
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/SamTemplateIndex.kt]: Ops, removed.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt (hunk 8 above)
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt]: Renamed.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt (hunk 9 above)
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/upload/LambdaLineMarker.kt]: Good point, I've added another method in the `LambdaHandlerResolver` to handle language level check.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt [hunk 15]
```diff
@@ -0,0 +1,42 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.cloudformation
+
+import software.aws.toolkits.jetbrains.services.cloudformation.IndexedResource.Companion.from
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: ?
